File: 98-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        return validate(root, null, null);
    }  
    
    private boolean validate(TreeNode root, Integer low, Integer high) {
        if(root == null) { return true; }
        if((low != null && root.val <= low) || (high != null && root.val >= high)) { return false; }
        
        return validate(root.left, low, root.val) && validate(root.right, root.val, high);
    }
    /*public boolean isValidBST(TreeNode root) {
        if(root == null) { return true; }
        
        Integer prev = null;
        Deque<TreeNode> stack = new ArrayDeque<>();
        TreeNode current = root;
        while(current != null || !stack.isEmpty()) {
            if(current != null) {
                stack.push(current);
                current = current.left;
            }
            else {
                 current = stack.pop();
                if(prev != null && current.val <= prev) {
                    return false;
                }
                prev = current.val;
                current = current.right;               
            }
        }
        
        return true;
    }*/
    
    Integer prev;
    /*public boolean isValidBST(TreeNode root) {
        prev = null;
        return inorder(root);
    }*/
    
    private boolean inorder(TreeNode root) {
        if(root == null) { return true; }
        
        if(!inorder(root.left)) { return false; }
        
        if(prev != null && root.val <= prev) {
            return false;
        }
        
        prev = root.val;
        return inorder(root.right);
    }
}

File: 98_Validate_Binary_Search_Tree.py
# 
A BST is valid only if every node fits into a valid value range.
I use DFS and pass down a (lower, upper) bound. For each node, 
its value must be strictly between those bounds. 
For the left child, the upper bound becomes the current node’s value; 
for the right child, the lower bound becomes the current value. 
If any node violates the bounds, I return false. If all nodes satisfy their ranges, it’s a valid BST.



class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def validate(node, lower, upper):
            if not node:
                return True
            
            # Check BST property: value must be in (lower, upper)
            if not (lower < node.val < upper):
                return False
            
            # Left: upper bound becomes node.val
            # Right: lower bound becomes node.val
            return validate(node.left, lower, node.val) and \
                   validate(node.right, node.val, upper)
        
        return validate(root, float("-inf"), float("inf"))
